### PR TITLE
Virtualize filenames as in-container paths from point of view of WDL command

### LINKED ISSUE #4514: Passing File arguments in an expression placeholder results in toil-wdl-runner failing to import those files
I previously thought this was something with how Toil converts strings into paths, but it looks like that functionality is good and that the cause of this error is something else.

When calling a WDL function that takes a File as an argument, putting the function call in an expression placeholder will result in an AssertionError that the file is not a local file. It looks like it happens on any function that takes a File as an argument, and passing a String normally works.

This occurs both when the file is defined inside the workflow and when the file is defined in the JSON input.

```wdl
workflow wf {
  input {}
  call t {
  }
}

task t {
  input {}
  File f = "/home/heaucques/Documents/toil-examples/test.wdl" # any file, error also happens if f is defined from the json
  command {
    echo ${size(f)} > output.txt
  }
}
```
```
[2023-06-28T15:25:08-0700] [MainThread] [E] [toil.wdl.wdltoil] Expression evaluation failed for command: 
	    echo ~{size(f)} > output.txt
	  
	Traceback (most recent call last):
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/Expr.py", line 123, in eval
	    ans = self._eval(env, stdlib)
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/Expr.py", line 1140, in _eval
	    return f(self, env, stdlib)
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/StdLib.py", line 232, in __call__
	    return self._call_eager(expr, [arg.eval(env, stdlib=stdlib) for arg in expr.arguments])
	  File "/home/heaucques/Documents/toil/src/toil/wdl/wdltoil.py", line 366, in _call_eager
	    total_size += os.path.getsize(self.stdlib._devirtualize_filename(uri))
	  File "/home/heaucques/Documents/toil/src/toil/wdl/wdltoil.py", line 452, in _devirtualize_filename
	    assert os.path.exists(result), f"Virtualized file {filename} looks like a local file but isn't!"
	AssertionError: Virtualized file /mnt/miniwdl_task_container/work/_miniwdl_inputs/0/test.wdl looks like a local file but isn't!
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1359)
┆Issue Number: TOIL-1359

### LINKED ISSUE #4515: toil-wdl-runner returns CommandFailed when write_json is in an expression placeholder
Calling write_json in the `command {}` section results in toil-wdl-runner returning a CommandFailed error
```wdl
workflow wf {
  input {}
  call t {}
  output {
    File out = t.out
  }
}

task t {
  input {}
  Map[String, String] map = {"k1": "v1", "k2": "v2", "k3": "v3"}
  command {
    cat ${write_json(map)} > output.txt
  }
  output {
    File out = "output.txt"
  }
}
```
```
[2023-06-28T15:58:54-0700] [MainThread] [I] [toil.wdl.wdltoil] Standard error at /tmp/314d8faf84675aff8c33e3e2f90d2189/48a7/d0f5/stderr.txt: cat: 'toilfile:36%3A0%3Afiles%2Ffor-job%2
Fkind-WDLTaskJob%2Finstance-_2rpv_9v%2Ffile-b61f80abe680450e840210e2d3ebe126%2Ftmpwnqn2gnv/tmpwnqn2gnv': No such file or directory
	
	[2023-06-28T15:58:54-0700] [MainThread] [I] [toil.wdl.wdltoil] Standard output at /tmp/314d8faf84675aff8c33e3e2f90d2189/48a7/d0f5/stdout.txt: 
	Traceback (most recent call last):
	  File "/home/heaucques/Documents/toil/src/toil/worker.py", line 403, in workerScript
	    job._runner(jobGraph=None, jobStore=jobStore, fileStore=fileStore, defer=defer)
	  File "/home/heaucques/Documents/toil/src/toil/job.py", line 2774, in _runner
	    returnValues = self._run(jobGraph=None, fileStore=fileStore)
	  File "/home/heaucques/Documents/toil/src/toil/job.py", line 2691, in _run
	    return self.run(fileStore)
	  File "/home/heaucques/Documents/toil/src/toil/wdl/wdltoil.py", line 1244, in run
	    task_container.run(miniwdl_logger, command_string)
	  File "/home/heaucques/Documents/toil/venv/lib/python3.10/site-packages/WDL/runtime/task_container.py", line 326, in run
	    raise CommandFailed(
	WDL.runtime.error.CommandFailed: task command failed with exit status 1
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1360)
┆Issue Number: TOIL-1360

### PR DESCRIPTION
**Pull request by adamnovak**
This should fix problems with WDL file manipulation functions not working right when called from command placeholder substitutions.

Now we have the "virtual" file name space be the inside of the container path space during WDL task commands, while the devirtualized path space is always the host side path space.

This should fix #4514 for files that aren't just converted from strings where that isn't really allowed, and should also fix #4515.

It doesn't really do anything for #4516, but is sort of relevant. With this, you *only* get the in-container path insode the command itself. If you go File to String elsewhere in the task, you get a URI.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * WDL commands now see in-container paths for files but devirtualize to out-of-container paths

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by adamnovak** [comment]
I tested this manually with:

```
cat >test.wdl <<'EOF'
version 1.0
workflow wf {
  input {}
  call t {
  }
  output {
    File out = t.out
  }
}

task t {
  input {}
  File f = write_lines(["hi"]) 
  command <<<
    echo ~{f} > output.txt
    echo ~{sep="," read_lines(f)} >>output.txt
    echo ~{size(f)} >>output.txt
    echo ~{write_lines(["new file in placeholder"])} >>output.txt 
  >>>
  output {
    File out = "output.txt"
  }
}
EOF
DOCKER_HOST=unix:///Users/anovak/.docker/run/docker.sock toil-wdl-runner test.wdl --logDebug --retryCount 0
```

The `DOCKER_HOST` part is just for me, to make sure Toil talks to Docker Desktop. at its non-standard location.

The output looks like this now (instead of crashing):

```
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/tmp4au6lvoy
hi
3.000000
/mnt/miniwdl_task_container/work/_miniwdl_inputs/0/tmpxbsc7t9s
```

I think this is right: `File`s coming into the command via bindings and via expressions evaluated in placeholders both produce in-container paths, but `File`s can still be operated on by the standard library functions.

**Comment by adamnovak** [comment]
Shloka said this works for her real workflow.

**Comment by stxue1** [comment]
This looks good to me. The fix runs right on my machine and it looks similar to how MiniWDL deals with placeholder expressions. A bunch of the conformance tests all run properly on my end as well.

File: src/toil/wdl/wdltoil.py [hunk 1]
```diff
@@ -470,6 +489,63 @@ def _virtualize_filename(self, filename: str) -> str:
...
+            # on the out-of-container paths.
+            raise RuntimeError(f"File {filename} is a URL but should already be an in-container-virtualized filename")
+        
+        # If this is a local path it will be in the container. Make sure we
+        # use the out-of-container equivalent.
+        result = self.container.host_path(filename)
+
+        if result is None:
+            # We really shouldn't have files in here that we didn't virtualize.
+            raise RuntimeError(f"File {filename} in container is not mounted from the host and can't be opened form the host")
```
Comment (DailyDreaming) [inline src/toil/wdl/wdltoil.py:528]: `opened form the host`

**Review by DailyDreaming — APPROVED** [review]
LGTM.  Minor comment.

**Pull request merged**